packages/app-desktop/gui/NoteEditor/utils/useFormNote.ts

```diff
@@ -189,13 +191,19 @@ export default function useFormNote(dependencies: HookDependencies) {
 
 			const focusSettingName = noteIsTodo ? 'newTodoFocus' : 'newNoteFocus';
 
-			requestAnimationFrame(() => {
+			autoFocusTimeout = setTimeout(() => requestAnimationFrame(() => {
 				if (Setting.value(focusSettingName) === 'title') {
-					if (titleInputRef.current) titleInputRef.current.focus();
+					if (titleInputRef.current) {
+						titleInputRef.current.focus();
+						focused = true;
+					}
 				} else {
-					if (editorRef.current) editorRef.current.execCommand({ name: 'editor.focus' });
+					if (editorRef.current) {
+						editorRef.current.execCommand({ name: 'editor.focus' });
+						focused = true;
+					}
```

Comment:
I mean how is that thing even working? You're pretty much crossing your fingers here and hoping that the two refs will be set after 50 milliseconds. Which in practice means it will randomly fail on slower computers. Not to mention you're never clearing the timeout if the effect is cancelled within less than 50 milliseconds. And as you said maybe you need the `focused` variable, maybe you don't and it seems you just sprinkle that variable here just in case. That's not how software development work.

You also ignored my previous comment where I mentioned using `setInterval`. We've been through I don't know how many revisions at this point and it's still not good.

Reply:
I agree, when I was writing this it felt like a really bad band aid solution rather than finding the source of why editorRef isn't defined in time. I forgot to mention that I didn't ignore the comment where you said to use `setInterval` but rather I kept the timeout because I felt like an interval would be more taxing on performance than was necessary so I hesitated on that. Should I continue with the interval to wait for the editorRef to be ready in that case or try to find out why editorRef isn't defined until later?

Reply:
So to clarify:

- Your previous solution using `setTimeout` was a bit more correct, but too complicated so it wasn't clear what was happening. Your solution was essentially simulating `setInterval` because you were resetting the timeout on each loop.
- That's why I suggested using `setInterval` to simplify the code, so that we can better understand it and be more confident it actually works.
- The way you would do this is like this:
    - Call `setInterval` 
    - At the top of the interval, check that editorRef AND titleInputRef are defined
    - If they are, set the focus
    - And then clear the interval
    - Get rid of this `requestAnimationFrame` call. It probably does nothing here and could introduce timing-related bugs.
- With this approach you can use a fast interval because it will keep trying until the condition is met

I hope that makes sense and please ask question is something is unclear. I think you're digging too deep in possibly unrelated code at the moment.

Reply:
That definitely clears things up, I removed the requestAnimationFrame and added the specified changes but currently this still doesn't work and lands us in the same situation as before. I committed what we have now so you can have a look at it. Having the clearInterval in the effect cleanup prevents it from auto focusing at all, but I know we need it there in the event that there is no editor or title ref to focus on, thus creating an infinite loop, so I left it there for now.

While doing a quick debugging, I temporarily removed the `clearInterval` so I can actually debug it and stepping through it shows its working as expected, leaving us with the same timing issue that I mentioned previously. I was hoping the other editor components would be unrelated but going through this more and more I have a feeling we might need to look into that and this might be a bigger issue than we thought. Should I add a check for the editorRef being defined in the execCommand functions? I'm worried if I add that it could cause some other issues if I add that, so maybe a limited number of checks to prevent an infinite loop there? (bear in mind this was first idea off the top of my head, there's likely a better way of checking this or I could be way overthinking this) What should I do in this case?

Reply:
What happens is a problem that's common with such complex effect that's doing a lot of unrelated things. Here the note loads, which triggers your interval. Then something changes, probably in `formNote`, which is a dependency of the effect. This unmount the effect and mount it again, however since formNote.id hasn't changed the effect is skipped (see condition at the top).

The solution to this is to refactor and move the autofocus code to its own effect, with its own conditions. Then you can better control when it will run and what happens when it is unmounted. So would you mind trying this and move your `handleAutoFocus` function to its own new effect? I think as condition you'll need to find out when a new note is loaded which you can probably figure out by checking formNote.id and previousNoteId

Reply:
Moving to a new effect does mitigate it and I found that it does work, but only most of the time, where VERY occasionally it still fails.